dronegen/windows.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2021 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import "path"
+
+const (
+	perBuildWorkspace   = `$Env:WORKSPACE_DIR/$Env:DRONE_BUILD_NUMBER`
+	windowsToolchainDir = perBuildWorkspace + `/toolchains`
+	perBuildTeleportSrc = perBuildWorkspace + "/go/src/github.com/gravitational/teleport"
+	perBuildWebappsSrc  = perBuildWorkspace + "/go/src/github.com/gravitational/webapps"
+
+	// Hardcoded tool versions that we would normally pull from the makefile,
+	// but unfortunately our Makefiles use too many POSIX-isms for us to use on
+	// Windows, so for now we will just say the versions we want here.
+
+	windowsNodeVersion = "16.13.2"
+)
+
+func newWindowsPipeline(name string) pipeline {
+	p := newExecPipeline(name)
+	p.Workspace.Path = path.Join("C:/Drone/Workspace", name)
+	p.Concurrency.Limit = 1
+	p.Platform = platform{OS: "windows", Arch: "amd64"}
+	return p
+}
+
+func windowsTagPipeline() pipeline {
+	p := newWindowsPipeline("build-native-windows-amd64")
+	p.Trigger = triggerTag
+	p.DependsOn = []string{"build-windows-amd64"}
+	p.Steps = []step{
+		cloneWindowsRepositoriesStep(p.Workspace.Path),
+		updateWindowsSubreposStep(p.Workspace.Path),
+		installWindowsNodeToolchainStep(p.Workspace.Path),
+		{
+			Name: "Fetch pre-built tsh",
+			Environment: map[string]value{
+				"WORKSPACE_DIR":         {raw: p.Workspace.Path},
+				"AWS_REGION":            {raw: "us-west-2"},
+				"AWS_S3_BUCKET":         {fromSecret: "AWS_S3_BUCKET"},
+				"AWS_ACCESS_KEY_ID":     {fromSecret: "AWS_ACCESS_KEY_ID"},
+				"AWS_SECRET_ACCESS_KEY": {fromSecret: "AWS_SECRET_ACCESS_KEY"},
+			},
+			Commands: []string{
+				`$Workspace = "` + perBuildWorkspace + `"`,
+				`$TeleportSrc = "` + perBuildTeleportSrc + `"`,
+				`$TeleportVersion=$Env:DRONE_TAG.TrimStart('v')`,
+				`$InputsDir="$Workspace/inputs"`,
+				`$DownloadDir="$Workspace/downloads"`,
+				`$TshZip="$InputsDir/teleport.zip"`,
```

Comment:
Where are you getting tsh from here? It's fine for the first iteration to fetch a pre-built tsh, but I think for consistency we should strive to build tsh as part of the build just like we do for Mac builds.

Reply:
Yes, that's the goal, but building `tsh` from scratch involves CGO, which in turn means getting the MinGW toolchain set up as well because CGo and MSVC don't appear to play nicely together. 

I was hoping to do that on a per-build basis, and so until then I'm pulling a pre-built `tsh` from dronestorage

Reply:
I've installed MinGW-w64 on the buildbox and `tsh.exe` is being built locally.